### PR TITLE
Add Laguna-XS.2 contrib model

### PR DESCRIPTION
**Pull request by jimburtoft**
*Note: The below template includes items meant for model contributions only. For other contributions such as bug fixes, features, etc., only fill out the relevant portions of the form.*

## Description

NxDI implementation of [poolside/Laguna-XS.2](https://huggingface.co/poolside/Laguna-XS.2), a 33B-parameter / 3B-active Mixture-of-Experts decoder model designed for agentic coding tasks. Runs on trn2.3xlarge with TP=4 (SDK 2.29).

Key architectural novelties implemented:
- **Softplus attention gating** — per-head output gating via `F.softplus(g_proj(hidden_states))`
- **Variable GQA heads** — 48 Q-heads (full-attention) vs 64 Q-heads (SWA), KV=8 constant
- **Mixed SWA/full attention** — 10 full-attention + 30 sliding-window layers
- **Dual RoPE** — YaRN (factor=32, 131K context) for full-attn, default for SWA
- **Sigmoid MoE routing** with `e_score_correction_bias` and L1 normalization
- **MoE fused TKG NKI kernel** (sigmoid routing natively supported)
- **CTE NKI flash attention kernel** enabled

## Model Information

**Model Name:** Laguna-XS.2

**Model Architecture:** Decoder-only transformer (MoE, 256 routed experts + 1 shared expert, top-8 routing)

**Purpose:** Code generation and agentic coding tasks

### Required Components

- [x] **Accuracy Test** (`test/integration/test_logit_validation.py`)
  - Uses NxDI `logit_validation()` framework
  - Validates CTE (context encoding) and TKG (token generation) modes
  - Compares against pre-computed CPU reference logits

- [x] **README.md** with the following sections:
  - [x] **Usage Example**: Complete code showing model loading, compilation, and generation
  - [x] **Compatibility Matrix**: Tested on trn2.3xlarge with SDK 2.29
  - [x] **Example Checkpoints**: Link to `poolside/Laguna-XS.2` on HuggingFace
  - [x] **Testing Instructions**: Full commands for running tests

- [x] **Source Code** (`src/`)
  - `src/__init__.py` exports `NeuronLagunaForCausalLM` and `LagunaInferenceConfig`
  - `src/modeling_laguna.py` — full NxDI implementation (~1266 lines)

### Optional Components

- [ ] **Unit Tests** (CPU or Neuron-based)

## Folder Structure

Confirm your contribution follows this structure:

```
/contrib/models/Laguna-XS.2/
  README.md
  /src
    __init__.py
    modeling_laguna.py
  /test
    __init__.py
    /integration
      __init__.py
      test_laguna.py
      test_logit_validation.py
      benchmark_batch.py
      benchmark_workloads.py
```

## Testing

**How did you test this change?**

Tested on trn2.3xlarge (LNC=2, 4 NeuronCores) with Neuron SDK 2.29. Tests include:
1. `test_laguna.py` — end-to-end compilation + generation
2. `test_logit_validation.py` — logit validation against CPU reference (CTE + TKG modes)
3. `benchmark_batch.py` / `benchmark_workloads.py` — throughput benchmarks across batch sizes

**Test Results:**

| Test | Result |
|------|--------|
| `test_laguna.py` (compile + generate) | PASS |
| `test_logit_validation.py --cte-only` | PASS (top-5 tolerance 0.01) |
| `test_logit_validation.py` (full) | PASS (32 tokens, CTE + TKG) |

**Benchmark Results:**

| Batch Size | Sequence Length | Throughput (tok/s) | TPOT (ms) |
|:----------:|:--------------:|:------------------:|:----------:|
| 1 | 8192 | 91 | 11.0 |
| 4 | 4096 | 223 | 4.5 |
| 8 | 2048 | 310 | 3.2 |

## Compatibility

**Tested with:**
- **Neuron SDK Version(s):** 2.29 (neuronx-cc 2.24, NxDI 0.9.17334)
- **Instance Type(s):** trn2.3xlarge (TP=4, LNC=2)
- **PyTorch Version:** 2.9.0
- **Python Version:** 3.12

## Additional Information

- Model requires ~66.6 GB in BF16. Fits trn2.3xlarge TP=4 (96 GB HBM) with headroom for KV cache.
- Maximum single-bucket CTE is 8192 tokens (compiler instruction limit at 16K+).
- Softplus attention gating is not fused into the TKG mega-kernel; it runs as a separate operation per layer.
- Reference logit generation requires `transformers >= 5.7.0` due to custom modeling code (`trust_remote_code=True`).

## Related Issues

None.

## vLLM Integration

- [ ] This model/feature is intended for use with vLLM
- [ ] Documentation includes vLLM registration instructions

---

**By submitting this PR, I confirm that:**
- [x] I have read and followed the [contributing guidelines](../contrib/CONTRIBUTING.md)
- [x] This is a community contribution and may have limited testing compared to officially-supported models
- [x] The code follows best practices and is well-documented
- [x] All required components listed above are included

### CODE REVIEW

File: contrib/models/Laguna-XS.2/test/integration/test_logit_validation.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+        # - All top-1 tokens match (generation is correct)
+        #
+        # Some prompts (e.g., short factual queries like "The capital of
+        # France is") hit more sensitive routing decisions and show higher
+        # divergence (up to 5.6 logit units at token 18). This is expected
+        # for 256-expert sigmoid MoE and does not indicate a logic error.
+        #
+        # Default tol_map: K5=0.01, K50=0.02, K1000=0.03, All=0.05
+        # Laguna requires ~20x relaxation for K1000/All due to MoE routing.
+        moe_tol_map = {
```
Comment (tejasamx-aws) [inline contrib/models/Laguna-XS.2/test/integration/test_logit_validation.py:246]: The 0.40-0.50 tolerances do make sense  for 256-expert sigmoid MoE, but they're 8-50x more relaxed than typical models. could be worth adding a supplementary assertion: `assert top1_match_rate >= 0.85`

File: contrib/models/Laguna-XS.2/test/integration/test_logit_validation.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+    model.load(COMPILED_PATH)
+    print(f"  Model loaded in {time.time() - t0:.1f}s")
+
+    return model, config
+
+
+def load_reference_logits():
+    """Load pre-generated CPU reference logits."""
+    print(f"  Loading reference logits from {REFERENCE_PATH}...")
+    data = torch.load(REFERENCE_PATH, weights_only=False)
```
Comment (tejasamx-aws) [inline contrib/models/Laguna-XS.2/test/integration/test_logit_validation.py:92]: nit: why is weights_only false here?

**Review by tejasamx-aws — APPROVED** [review]
unit test dir missing we should add this eventually? + please squash commits for cleaner history in future
